### PR TITLE
Add `fn Atomic::from_impl`, a `const fn` way to create an `Atomic`, but from the underlying atomic impl directly

### PR DESCRIPTION
**Pull request by kkysen**
I'm trying to have a `static` `Atomic<T>`, but I can't initialize it as there's no `const fn` way to create an `Atomic<T>`, since `Atom` is a `trait` and `const fn`s on `trait`s aren't supported yet.  So I added this `const fn` method `from_impl` that takes the underlying atomic impl instead of a `T`, relying on the user to call a concrete `const fn` to get the atomic type from `T`.

As an example, I could do this:

```rust
use atomig::Atom;
use atomig::Atomic;
use bitflags::bitflags;
use std::sync::atomic::AtomicU8;

#[derive(Atom)]
pub struct Flags(u8);

bitflags! {
    impl Flags: u8 {}
}

static FLAGS: Atomic<Flags> = Atomic::from_impl(AtomicU8::new(Flags::empty().bits()));
```

Previously, this `PrimitiveAtom::Impl` associated type was `#[doc(hidden)]`, but now that it is part of the public API through `Atomic::from_impl`, it must be public.

### CODE REVIEW

**Comment by LukasKalbertodt** [comment]
Thanks for the PR and bringing attention to this issue (of not having a good way to initialize `static` `Atomic`s). I looked into this as well now and huff, there isn't really a good solution I suppose :/ 

I quite like your idea. But it has one disadvantage: it exposes the associated type `Impl` which is currently implementation detail and marked as `#[doc(hidden)]`. If we go with your solution, we would have "promote" that to be part of the public API of atomig. Also: what do you think of `from_impl` as name? I think the `new_` prefix doesn't really serve a purpose and `impl` fits the associated type better than `primitive`.

I had two other ideas. I tried to avoid that users have to create the `Atomic*` type themselves as that's quite a bit of boilerplate code. 
- First I thought I could use `mem::transmute` to convert from `PrimitiveAtom` to its `::Impl`. From reading the std docs it seems like this is indeed safe for all `Atomic*` types. However, `transmute` in generic contexts fails as it cannot guarantee that source and destination type indeed have the same size. 
- My only other idea was to just avoid generic contexts and just add a bunch of `const fn from_u8()` methods to `Atomic`, one for each primitive atomic. That works, with the obvious disadvantage of a somewhat bloated API.

So: what do you think? Should we go your route (with the slightly more verbose call site and requirement to expose more of the `PrimitiveAtom` trait) or go the route with the individual constructor methods (each a bit nicer to use, but having 12 of them). Or both?

**Comment by kkysen** [comment]
> But it has one disadvantage: it exposes the associated type `Impl` which is currently implementation detail and marked as `#[doc(hidden)]`. If we go with your solution, we would have "promote" that to be part of the public API of atomig.

> Also: what do you think of `from_impl` as name? I think the `new_` prefix doesn't really serve a purpose and `impl` fits the associated type better than `primitive`.

That would also work pretty well I think.  I wasn't really sure about the naming at all.

> I had two other ideas. I tried to avoid that users have to create the `Atomic*` type themselves as that's quite a bit of boilerplate code.

I don't think the boilerplate for this is too bad, and people are already used to the atomic types, and of having to write unwieldy code in `const` contexts (e.x. no `for` loops).  And compared to the `from_uN` methods, it's similar level of boilerplate.

> * First I thought I could use `mem::transmute` to convert from `PrimitiveAtom` to its `::Impl`. From reading the std docs it seems like this is indeed safe for all `Atomic*` types. However, `transmute` in generic contexts fails as it cannot guarantee that source and destination type indeed have the same size.
> * My only other idea was to just avoid generic contexts and just add a bunch of `const fn from_u8()` methods to `Atomic`, one for each primitive atomic. That works, with the obvious disadvantage of a somewhat bloated API.
> 
> So: what do you think? Should we go your route (with the slightly more verbose call site and requirement to expose more of the `PrimitiveAtom` trait) or go the route with the individual constructor methods (each a bit nicer to use, but having 12 of them). Or both?

That said, I didn't know that `Impl` is hidden, so if that's a significant concern, then your suggestion is pretty good, too.  I'm not sure how significant it would be to make that type part of the public API.

Overall, both ways are fine I think.  Neither is ideal, but there is no ideal solution until we get `const fn`s in `trait`s.  So I'd say it's up to you.  Both seem good to me; I'm just not sure how to weigh some of these API considerations.  I did wish a const generic `uN` type existed.  That could've helped, I think.

**Comment by LukasKalbertodt** [comment]
> I don't think the boilerplate for this is too bad, and people are already used to the atomic types, and of having to write unwieldy code in `const` contexts (e.x. no `for` loops). And compared to the `from_uN` methods, it's similar level of boilerplate.

Fair points! Ok lets go with your suggestion then. Exposing the type in the public API shouldn't be a problem at all. Ok so please:
- Rename the method to `from_impl`
- Remove the `#[doc(hidden)]` from `PrimitiveAtom::Impl`
- Change the `impls` module doc comment:
  > Traits for abstracting over `std` atomics. Mostly hidden implementation detail.
  >
  > Most items of these traits are hidden and not part of the public API of this library. You cannot implement these traits yourself.
- I would reword the first sentence of the new method's docs: "Creates a new atomic value from the underlying `Atomic*` type from `std`". The reasoning for that is explained in the next paragraph you wrote.

**Comment by kkysen** [comment]
I updated it with all of those changes now.

Also, I was curious, `rustfmt` formats a bunch of things very differently for me, which I avoided including with this PR.  Are you using different formatting settings somewhere?

**Review by LukasKalbertodt — APPROVED** [review]
Thank you!

> Also, I was curious, `rustfmt` formats a bunch of things very differently for me, which I avoided including with this PR. Are you using different formatting settings somewhere?

I don't use rustfmt at all, as there are couple of choices that I don't agree with and that aren't configurable (on stable). I wasn't really aware of how many people use rustfmt to autoformat everything (I had similar interactions with other people on other PRs in other repos in the last few months...). Anyway, thanks for not including these stray formatting changes in this PR!

**Comment by LukasKalbertodt** [comment]
Oh woops, just one line-length violation to be fixed.

**Comment by kkysen** [comment]
> Oh woops, just one line-length violation to be fixed.

Fixed!

> I don't use rustfmt at all, as there are couple of choices that I don't agree with and that aren't configurable (on stable). I wasn't really aware of how many people use rustfmt to autoformat everything (I had similar interactions with other people on other PRs in other repos in the last few months...). Anyway, thanks for not including these stray formatting changes in this PR!

Ah.  There are some things about `rustfmt`, I don't like either and I wish it had some more customization, but I've found it harder to fight against it, and there was no great alternative.

Also, could you let me know when you make a new release with this and #12 in it?  Thanks!

**Comment by LukasKalbertodt** [comment]
This and #12 is now released as 0.4.1
